### PR TITLE
Atom type generator

### PR DESCRIPTION
**Pull request by sblackburn86**
Adding the functionalities to generate atom types in parallel to the relative coordinates in the Langevin solver. The other generators are reworked to deal with AXL objects.

The constrained Langevin solver is not fully adapted: we cannot constraint atom types. This would go in a different PR later.

I had to move a function from the D3PMLoss to d3pm_utils - unit tests followed.

The bulk of the changes are in the generators/ folder, everything else should be fairly simple changes to accomodate the new AXL-ized generators.

### CODE REVIEW

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py [hunk 1]
```diff
@@ -23,38 +23,102 @@ def class_index_to_onehot(index: torch.Tensor, num_classes: int) -> torch.Tensor
...
+    q_matrices: torch.Tensor,
+    q_bar_matrices: torch.Tensor,
+    q_bar_tm1_matrices: torch.Tensor,
+    small_epsilon: float,
+    probability_at_zeroth_timestep_are_onehot: bool = True,
+) -> torch.Tensor:
+    r"""Compute :math:`P(a_{t-1} | a_t, a_0)`, for given probability distribution a_0 and a_t.
+
+    .. math::
+        P(a_{t-1} | a_t, a0_like) = (a_0^T \cdot \bar{Q}_{t-1} \cdot a_{t-1}) (a_{t-1}^T \cdot Q_t \cdot a_t) /
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:76]: `a0_like` and `a_0` should be consistent. Perhaps `gamma_0`?

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py [hunk 2]
```diff
@@ -23,38 +23,102 @@ def class_index_to_onehot(index: torch.Tensor, num_classes: int) -> torch.Tensor
...
+                                        (a_0^T \cdot \bar{Q}_{t} \cdot a_t)
+
+    Args:
+        probability_at_zeroth_timestep: a probability representation of a class type (one-hot
+            distribution or normalized distribution), as a tensor with dimension
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:81]: Indicate which symbol represents this in the math equation above.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py [hunk 3]
```diff
@@ -23,38 +23,102 @@ def class_index_to_onehot(index: torch.Tensor, num_classes: int) -> torch.Tensor
...
+            [batch_size, number_of_atoms, num_classes]
+        one_hot_probability_at_current_timestep: a one-hot representation of a class type at current time step, as a
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:83]: Indicate which symbol represents this in the math equation above.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py [hunk 4]
```diff
@@ -23,38 +23,102 @@ def class_index_to_onehot(index: torch.Tensor, num_classes: int) -> torch.Tensor
...
+            tensor with dimension [batch_size, number_of_atoms, num_classes]
+         q_matrices: transition matrices at current time step :math:`{Q}_{t}` of dimension
+            [batch_size, number_of_atoms, num_classes, num_classes].
+        q_bar_matrices: one-shot transition matrices at current time step :math:`\bar{Q}_{t}` of dimension
+            [batch_size, number_of_atoms, num_classes, num_classes].
+        q_bar_tm1_matrices: one-shot transition matrices at previous time step :math:`\bar{Q}_{t-1}` of dimension
+            [batch_size, number_of_atoms, num_classes, num_classes].
+        small_epsilon: minimum value for the denominator, to avoid division by zero.
+        probability_at_zeroth_timestep_are_onehot: if True, assume the probability_at_zeroth_timestep sum to 1.
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:92]: I think you mean `probability_at_zeroth_timestep_are_logits`.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py [hunk 5]
```diff
@@ -23,38 +23,102 @@ def class_index_to_onehot(index: torch.Tensor, num_classes: int) -> torch.Tensor
...
+
+
+def get_probability_at_previous_time_step(
+    probability_at_zeroth_timestep: torch.Tensor,
+    one_hot_probability_at_current_timestep: torch.Tensor,
+    q_matrices: torch.Tensor,
+    q_bar_matrices: torch.Tensor,
+    q_bar_tm1_matrices: torch.Tensor,
+    small_epsilon: float,
+    probability_at_zeroth_timestep_are_onehot: bool = True,
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:71]: I think you mean `probability_at_zeroth_timestep_are_logits`.

File: src/diffusion_for_multi_scale_molecular_dynamics/loss/atom_type_loss_calculator.py [hunk 6]
```diff
@@ -190,14 +153,15 @@ def get_p_atm1_given_at(
...
-                                                                   small_epsilon)
-
+        p_atm1_at = get_probability_at_previous_time_step(
+            probability_at_zeroth_timestep=predicted_logits,
+            one_hot_probability_at_current_timestep=one_hot_at,
+            q_matrices=q_matrices,
+            q_bar_matrices=q_bar_matrices,
+            q_bar_tm1_matrices=q_bar_tm1_matrices,
+            small_epsilon=small_epsilon,
+            probability_at_zeroth_timestep_are_onehot=False,
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/loss/atom_type_loss_calculator.py:163]: I think it would be more clear if this was ` probability_at_zeroth_timestep_are_logits=True`

File: tests/utils/test_sample_trajectory.py [hunk 7]
```diff
@@ -210,16 +354,31 @@ def test_record_corrector(
...
-    )
+    for axl_field, axl_name in AXL_NAME_DICT.items():
+        corrector_i = torch.stack(
+            sample_trajectory.data[f"corrector_{axl_name}_i"], dim=0
+        )
+        target_corrector_i = torch.stack(
+            [getattr(axl, axl_field) for axl in list_axl_i_corr], dim=0
+        )
+        torch.testing.assert_close(corrector_i, target_corrector_i)
+        corrector_corrected_im1 = torch.stack(
```
Comment (rousseab) [inline tests/utils/test_sample_trajectory.py:365]: Shouldn't this be `corrector_corrected_im1` -> `  corrector_corrected_i`? Why `im1`?

Same thing for the variables below.

File: src/diffusion_for_multi_scale_molecular_dynamics/generators/sde_position_generator.py [hunk 8]
```diff
@@ -125,9 +131,9 @@ def f(
         """
         diffusion_time = self._get_diffusion_time(sde_time)
 
-        sigma_normalized_scores = self.get_sigma_normalized_score(
-            diffusion_time, flat_relative_coordinates
-        )
+        sigma_normalized_scores = self.get_model_predictions(
+            diffusion_time, flat_relative_coordinates, self.atom_types
+        ).X  # we are only using the sigma normalized score forthe relative coordinates diffusion
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/generators/sde_position_generator.py:136]: `forthe` -> `for the`

File: src/diffusion_for_multi_scale_molecular_dynamics/generators/langevin_generator.py [hunk 9]
```diff
@@ -41,39 +46,63 @@ def __init__(
...
         self,
-        x: torch.Tensor,
+        composition: AXL,
         time: float,
-        noise: float,
-        unit_cell: torch.Tensor,
+        sigma_noise: float,
+        unit_cell: torch.Tensor,  # TODO replace with AXL-L
         cartesian_forces: torch.Tensor,
     ) -> torch.Tensor:
```
Comment (rousseab) [inline src/diffusion_for_multi_scale_molecular_dynamics/generators/langevin_generator.py:96]: The return type is now and `AXL`.

**Review by rousseab — CHANGES_REQUESTED** [review]
This looks real good! I just have a few suggestions / typo fixes.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py (hunk 1 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py]: This is done.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py (hunk 2 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py:82]: done.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py (hunk 5 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py]: done.

File: src/diffusion_for_multi_scale_molecular_dynamics/generators/sde_position_generator.py (hunk 8 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/generators/sde_position_generator.py]: done.

File: src/diffusion_for_multi_scale_molecular_dynamics/generators/langevin_generator.py (hunk 9 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/generators/langevin_generator.py]: done.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py (hunk 3 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py]: done.

File: src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py (hunk 4 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py]: Done.

File: src/diffusion_for_multi_scale_molecular_dynamics/loss/atom_type_loss_calculator.py (hunk 6 above)
Comment (rousseab) [reply src/diffusion_for_multi_scale_molecular_dynamics/loss/atom_type_loss_calculator.py]: Done.

File: tests/utils/test_sample_trajectory.py (hunk 7 above)
Comment (rousseab) [reply tests/utils/test_sample_trajectory.py]: good enough.

**Review by rousseab — APPROVED** [review]
LGTM!
